tools/syz-trace2syz/proggen/proggen.go

```diff
@@ -0,0 +1,613 @@
+)
+
+type returnCache map[resourceDescription]prog.Arg
+
+func newRCache() returnCache {
+	return make(map[resourceDescription]prog.Arg)
+}
+
+func (r *returnCache) buildKey(syzType prog.Type) string {
+	switch a := syzType.(type) {
+	case *prog.ResourceType:
+		return "ResourceType-" + a.Desc.Kind[0]
+	default:
+		log.Fatalf("Caching non resource type")
+	}
+	return ""
+}
+
+func (r *returnCache) cache(syzType prog.Type, traceType parser.IrType, arg prog.Arg) {
+	log.Logf(2, "Caching resource type: %s, val: %s", r.buildKey(syzType), traceType.String())
+	resDesc := resourceDescription{
+		Type: r.buildKey(syzType),
+		Val:  traceType.String(),
+	}
+	(*r)[resDesc] = arg
+}
+
+func (r *returnCache) get(syzType prog.Type, traceType parser.IrType) prog.Arg {
+	log.Logf(2, "Fetching resource type: %s, val: %s", r.buildKey(syzType), traceType.String())
+	resDesc := resourceDescription{
+		Type: r.buildKey(syzType),
+		Val:  traceType.String(),
+	}
+	if arg, ok := (*r)[resDesc]; ok {
+		if arg != nil {
+			log.Logf(2, "Cache hit for resource type: %s, val: %s", r.buildKey(syzType), traceType.String())
+			return arg
+		}
+	}
+	return nil
+}
+
+func (r *returnCache) findMatchingType(syzType prog.Type, traceType parser.IrType) prog.Arg {
+	key := r.buildKey(syzType)
+	for k := range *r {
+		if key == k.Type {
+			return (*r)[k]
+		}
+	}
+	return nil
+}
+
+type resourceDescription struct {
+	Type string
+	Val  string
+}
+
+// Context stores metadata related to a syzkaller program
+// Currently we are embedding the State object within the Context.
+// We should probably merge the two objects
+type Context struct {
+	ReturnCache       returnCache
+	Prog              *prog.Prog
+	CurrentStraceCall *parser.Syscall
+	CurrentSyzCall    *prog.Call
+	CurrentStraceArg  parser.IrType
+	Target            *prog.Target
+	Tracker           *memoryTracker
+	CallToCover       map[*prog.Call][]uint64
+	Call2Variant      *CallVariantMap
+	DependsOn         map[*prog.Call]map[*prog.Call]int
+}
+
+func newContext(target *prog.Target, variantMap *CallVariantMap) (ctx *Context) {
+	ctx = &Context{}
+	ctx.ReturnCache = newRCache()
+	ctx.CurrentStraceCall = nil
+	ctx.Tracker = newTracker()
+	ctx.CurrentStraceArg = nil
+	ctx.Target = target
+	ctx.CallToCover = make(map[*prog.Call][]uint64)
+	ctx.Call2Variant = variantMap
+	ctx.DependsOn = make(map[*prog.Call]map[*prog.Call]int)
+	return
+}
+
+// FillOutMemory determines how much memory to allocate for arguments in a program
+// And generates an mmap c to do the allocation.This mmap is prepended to prog.Calls
+func (ctx *Context) FillOutMemory() error {
+	err := ctx.Tracker.fillOutMemory(ctx.Prog)
+	if err != nil {
+		return err
+	}
+	totalMemory := ctx.Tracker.getTotalMemoryAllocations(ctx.Prog)
+	log.Logf(2, "Total memory for program is: %d", totalMemory)
+	if totalMemory == 0 {
+		log.Logf(1, "Program requires no mmaps. Total memory: %d", totalMemory)
+		return nil
+	}
+	mmapCall := ctx.Target.MakeMmap(0, totalMemory)
```

Comment:
This is not necessary now, all memory is implicitly mapped. No need to emit explicit mmap call.

Reply:
ack